### PR TITLE
design-docs: describe process and add examples

### PR DESCRIPTION
**Pull request by athre0z**
This documents the design document process that we've been using to develop the profiling agent in the past.

For the most part I'm essentially just publishing the documentation of our previously internal process at optimyze/Elastic. Nothing of what I'm proposing here is set in stone, and I'd like to invite the new contributors from other companies to comment on it here if you'd like to change anything about it!

To provide some additional context of what a design document created with this process could look like, I also included two of our old profiling agent design documents. Many of the links had to be replaced with `#` placeholders because they refer to old git history that isn't available in the open-sourced repository, but I think they should still provide a good general idea.

### CODE REVIEW

**Review by christos68k — APPROVED** [review]
Thanks!

File: design-docs/README.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+### When to write a design document?
+
+Invoking the design document process can be beneficial if
+
+- you are proposing changes that have significant impact on multiple
+  sub-components of the profiling agent.
+- you are proposing a major addition and want to make sure that it will
+  actually be accepted upstream the way that you plan to implement it without
+  writing the code first (and thus risking need for major reworks later).
+- your proposed changes require a signficant amount of context to understand
```
Comment (rockdaboot) [inline design-docs/README.md:16]:
```suggestion
- your proposed changes require a significant amount of context to understand
```

File: design-docs/00000-example-docs/04359-remove-bpf-hashing/README.md [hunk 2]
```diff
@@ -0,0 +1,736 @@
...
+identifiers.
+
+**Validated in prototype:** no\
+**Independent issue:** no (unless we start without a trace translation cache)
+
+#### Hash algorithm
+
+BPF is currently using a hand-rolled, custom hashing scheme. The scheme was
+chosen primarily for simplicity and is likely neither the fastest, nor the most
+collision resistent function imaginable. We already have XXH3 as a dependency
```
Comment (rockdaboot) [inline design-docs/00000-example-docs/04359-remove-bpf-hashing/README.md:664]:
```suggestion
collision resistant function imaginable. We already have XXH3 as a dependency
```

**Review by rockdaboot — APPROVED** [review]

**Comment by felixge** [comment]
Sorry for the delay, I'm catching up on a backlog. This LGTM. I'm double checking with my colleagues, but unless they see any issues I'll approve on Monday.

File: design-docs/README.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    properly view rendered documents behind multiple clicks, so adding a direct
+    link spares others some work. After creating the PR, this link can be
+    obtained by clicking "Files changed", searching the `README.md` file,
+    clicking the three dots on the right, selecting "View file" and then
+    switching the current view to track your branch via the drop-down to the
+    left of the file path. The last step ensures that the link automatically
+    tracks updates when you push fixup commits later.
+- Rename the directory according to the PR ID that was allocated, e.g. if the
+  allocated PR ID is 1234 then `00000-my-doc-title` becomes
+  `00000-my-doc-title`
```
Comment (Gandem) [inline design-docs/README.md:53]: Small typo 🙏 - I assume the intent was:

```suggestion
  `01234-my-doc-title`
```

**Review by felixge — APPROVED** [review]

**Pull request merged**